### DISCUSSION TITLE
Lesson 8: Help needed to solve problem

### DISCUSSION THREAD

**Discussion by npa528** [Cyfrin/foundry-full-course-cu#102] [General]
As others pointed out in other discussion, (convert the IPFS to HTTP link through gateway), I was transferred to a page asking to connect with Metamask and just type my Twitter handle on textfiled and press button "solve"

![Challenge8](https://github.com/Cyfrin/foundry-full-course-f23/assets/5271051/9c5c6eba-3862-4f30-9e94-c34abf55ae63)

I did it but I get the error below:
```
MetaMask - RPC Error: execution reverted 
Object { code: -32000, message: "execution reverted" }
[inpage.js:1:78272](moz-extension://8dd80d6b-15f5-4fdf-b0fc-bf088e6fd1dc/inpage.js)
Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"code":-32000,"message":"execution reverted"}, method="estimateGas", transaction={"from":"0x05E3Bd725678352E24d79D2266e12129EdB69474","to":"0x76B50696B8EFFCA6Ee6Da7F6471110F334536321","data":"0x29a30a7000000000000000000000000000000000000000000000000000000000075bcd15000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000086368616d64657631000000000000000000000000000000000000000000000000","accessList":null}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.1.2)
```

I thought the challenge was to convert IPFS to HTTP, I cannot figure out what else to do to solve challenge.

I inspect code, I checked the TO address in console (0x76B50696B8EFFCA6Ee6Da7F6471110F334536321) but I cannot find what else to do to pass this challenge. 
Any hints?

**Comment by PatrickAlphaC** [comment]
Are you on the right network?

**Reply by npa528** [reply]
Yes, I am on Sepolia (11155111)

**Comment by alymurtazamemon** [comment]
@npa528 Hint: Take the contract address, read the ABI and send the transaction using the foundry `cast` command. Also, you need to find the secret code too!!!

**Reply by npa528** [reply]
Are you sure you are talking about Challenge 8 (at the end of Lesson 8)? I updated my initial post to upload image of the Challenge 8.

**Reply by alymurtazamemon** [reply]
@npa528 Yes!

**Reply by 0xS7AN** [reply]
I get the same error. Also tried to call it via foundry's `cast send` command and still get the same error back.

**Reply by palmcivet7** [reply]
I am also getting a similar error. I have tried:
- resetting Metamask account
- using a different browser
- using a different wallet
- using a different computer
- foundry cast send

When attempting with Foundry I get one of these two errors depending on how much value I set the `--gas-limit` and `--gas-price` in my `cast send` command:
```
code: -32000, message: insufficient funds for gas * price + value
```
or
```
code: -32000, message: intrinsic gas too low
```

When attempting with the IPFS page's solve button, the console logs the following before the subsequent error:
```
Solving...
Working on 42161
```
```
code -32603, message: 'Internal JSON-RPC error.'
```
The same results are occurring with both Arbitrum and Sepolia.
I opened a ticket with Metamask technical support and they told me to check the Consensys discord or make a thread on Stackexchange, but the issue has been asked in both of those places with no definitive answer.

**Comment by npa528** [comment]
I finally solved this challenge. I am very frustrated with whoever engineered this Challenge. It needs skills outside of the knowledge areas covered in this tutorial.

I can just post the solution but I guess moderators will not allow this. 

In order to solve it I had to dig a lot to sepolia.etherscan. The address that is returned from the IPFS page (0x76B50696B8EFFCA6Ee6Da7F6471110F334536321) is not the contract that you finally need to call!!!

From above contract (0x76B5...6321), I had to find 
  1) the contract creator, see what other contracts he had deployed, 
  2) find the one which looked suspicious (no name, not verified) based on the idea that it should be right after "**Create: LessonSeven**"
      since challenge LessonEight is missing from this list.

![Challenge8](https://github.com/Cyfrin/foundry-full-course-f23/assets/5271051/c172e79f-0419-44f7-8eae-23115144dcee)

This contract: 0xf215a0b6DD88d6029b5385D6fab51968337E963D
So THIS IS the contract that you need to call.

Now when you get there (in contract 0xf215...963D), I had to check latest transactions and especially click for "more details". 
I saw there 1-2 people who already solved this challenge and I blindly try to replicate their call.

Didn't know how or what i was doing, i just try to replicate what they did using "cast send" and putting the right values. 

If i had to rate this challenge i would put 0/5 stars. Literally I feel I wasted my time. Very poorly designed, I wish somebody could elaborate like what they had in mind making this puzzle so difficult for no reason....

I also would like to know a more "programmatically" way to solve this, and not digging like sleuth manually as I did

I hope this helps a bit.
PS. I am OK to send the solution to whoever wants it.

**Reply by 0xS7AN** [reply]
That helped, thanks.

**Reply by palmcivet7** [reply]
https://arbiscan.io/address/0x33ee14fb8816c92fe401165330bbe29706942183 This is the arbitrum contract, it is also different to the one in the constants file on IPFS, which is the contract for minting the NFTs.
Although I'd prefer not to get the answer outright, can you elaborate on how you solved it with `cast send` please? I'm still getting the same errors as above when I use the following:
```
cast send --rpc-url $ARBITRUM_RPC_URL --private-key $PRIVATE_KEY --gas-limit 50000 --gas-price 20gwei 0x33ee14Fb8816c92fe401165330bbE29706942183 "solveChallenge(uint256,string)" 123456789 “@palmcivetdev”
```

**Reply by npa528** [reply]
@palmcivet7 ,
In your case I would try your command without "--gas-limit" and "--gas-price" parameters.
And also try to remove quotes from twitterHandle string just in case. Like this:
```
cast send 0x33ee14Fb8816c92fe401165330bbE29706942183  "solveChallenge(uint256,string)" 123456789 palmcivetdev --rpc-url $ARBITRUM_RPC_URL --private-key $PRIVATE_KEY
```

and don't forget to "source .env" before you "cast send"

**Reply by palmcivet7** [reply]
Thank you for the help. Unfortunately I'm getting another gas related error:
```
(code: -32000, message: gas required exceeds allowance (468750), data: None)
```
Maybe I should put more ETH in my wallet. I've got a few dollars worth so I would've thought it'd be fine because all the other challenges cost around 15c.

**Reply by alymurtazamemon** [reply]
@npa528 I cannot spoil this challenge, but if you had understood the topics taught in this tutorial then you did not have said this!

You do not need to find deployers' transactions to find the contract address for this challenge's contract. 

Let us respect the one who created this because after all he put huge efforts into these challenges and courses and provided us with them free of cost.

**Reply by PatrickAlphaC** [reply]
Hmm... Thanks for the feedback on this though. It was supposed to be an easy one. Perhaps it didn't correctly swap to the sepolia address correctly... Truth be told I only tested it on arbitrum.

**Reply by PatrickAlphaC** [reply]
Apologies on the frustrations here... I'm OK to spoil this challenge because it's not supposed to be this difficult. 

But at the same time... @npa528 you used the skills we taught you to go above and beyond to solve the challenge!! Well done!

Often, doing really odd things as you did is the secret to an exploit, I hope to see you become a smart contract auditor later in your career. That initiative is what it takes to get there.

So, feel free to spoil this challenge.

**Reply by palmcivet7** [reply]
The `cast send` solution I posted above worked fine in the end after I funded my account with more ETH.

**Reply by john-lawniczak** [reply]
Kind of a dope way to solve this ngl.

**Reply by sharonphiliplima** [reply]
Kudos, I solved the challenge in the same way with a hint from an online friend. I have no idea why the solution is such a magic number ^^ 
Anyway, auditing would seem fun it seems!
